datapath.go

```diff
@@ -1,65 +1,139 @@
-	"bytes"
+	"bufio"
 	"encoding/binary"
 	"fmt"
-	"net"
-
 	"github.com/nutanix/gofc/ofprotocol/ofp13"
+	"net"
+	"sync/atomic"
 )
 
 // datapath
 type Datapath struct {
-	buffer     chan *bytes.Buffer
-	conn       *net.TCPConn
+	done       chan bool
+	conn       net.Conn
 	datapathId uint64
 	sendBuffer chan *ofp13.OFMessage
+	readBuffer chan []byte
 	ofpversion string
 	ports      int
+	server     *OFController
+	inShutdown int32
+	onClose    []func(dp *Datapath)
 }
 
 /**
  * ctor
  */
-func NewDatapath(conn *net.TCPConn) *Datapath {
+func NewDatapath(conn net.Conn, s *OFController) *Datapath {
 	dp := new(Datapath)
 	dp.sendBuffer = make(chan *ofp13.OFMessage, 10)
+	dp.readBuffer = make(chan []byte, 10)
+	dp.done = make(chan bool, 2)
 	dp.conn = conn
+	dp.server = s
 	return dp
 }
 
+func (dp *Datapath) Shutdown() {
+	atomic.StoreInt32(&dp.inShutdown, 1)
+	close(dp.readBuffer)
+	<-dp.done
+	close(dp.sendBuffer)
+	<-dp.done
+	close(dp.done)
+	dp.Close()
+}
+
+func (dp *Datapath) RegisterOnClose(f func(dp *Datapath)) {
+	if atomic.LoadInt32(&dp.inShutdown) == 1 {
+		return
+	}
+	dp.onClose = append(dp.onClose, f)
+}
+
+func (dp *Datapath) Close() {
+	atomic.StoreInt32(&dp.inShutdown, 1)
+	if err := dp.conn.Close(); err != nil {
+		fmt.Printf("%v", err)
+	}
+	for _, f := range dp.onClose {
+		f(dp)
+	}
+}
+
 func (dp *Datapath) sendLoop() {
+	defer func() { dp.done <- true }()
 	for {
 		// wait channel
-		msg := <-(dp.sendBuffer)
+		msg, more := <-(dp.sendBuffer)
 		// serialize data
+		if !more {
+			return
+		}
 		byteData := (*msg).Serialize()
 		_, err := dp.conn.Write(byteData)
 		if err != nil {
 			fmt.Println("failed to write conn")
 			fmt.Println(err)
+			dp.Close()
 			return
 		}
 	}
 }
 
-func (dp *Datapath) recvLoop() {
-	buf := make([]byte, 1024*64)
+func (dp *Datapath) parseLoop() {
+	defer func() { dp.done <- true }()
 	for {
-		// read
-		size, err := dp.conn.Read(buf)
-		if err != nil {
-			fmt.Println("failed to read conn")
-			fmt.Println(err)
+		buf, more := <-dp.readBuffer
+		if !more {
 			return
 		}
+		dp.handlePacket(buf)
+	}
+}
 
-		// tmp := make([]byte, 2048)
-		for i := 0; i < size; {
-			msgLen := binary.BigEndian.Uint16(buf[i+2:])
-			dp.handlePacket(buf[i : i+(int)(msgLen)])
-			i += (int)(msgLen)
+func (dp *Datapath) recvLoop() {
+	go dp.parseLoop()
+	// for more information see https://www.opennetworking.org/wp-content/uploads/2014/10/openflow-spec-v1.3.0.pdf
+	const kHeaderSize = 8
+	reader := bufio.NewReader(dp.conn)
+	for {
+		if atomic.LoadInt32(&dp.inShutdown) == 1 {
+			return
+		}
+		buf := make([]byte, 1024*64)
+		for i := 0; i < kHeaderSize; i++ {
+			b, err := reader.ReadByte() // read len prefix and len
+			if err != nil {
+				if atomic.LoadInt32(&dp.inShutdown) != 1 {
+					fmt.Println("failed to read conn")
+					fmt.Println(err)
+					dp.Close()
+				}
+				return
+			}
+			buf[i] = b
+		}
+		msgLen := (int)(binary.BigEndian.Uint16(buf[2:]))
+		bufPos := kHeaderSize
+		for bufPos < msgLen {
```

Comment:
we can create a separate buf here with size msgLen

Reply:
That was fixed with use one receiver buffer and copy data from slice to new buffer before send in to receive channel.